### DISCUSSION TITLE
`Mapper.primary_key` + Polymorphic Inheritance is not working as expected

### DISCUSSION THREAD

**Discussion by AlbertMitjans** [sqlalchemy/sqlalchemy#12157] [Usage Questions]
I will leave below a coding example of the problem I am facing:
```python3
from uuid import UUID, uuid4

from sqlalchemy import ForeignKey, String, inspect
from sqlalchemy.dialects.postgresql import UUID as PSQL_UUID
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class Interlocutor(Base):
    __tablename__ = "interlocutors"
    __mapper_args__ = {"polymorphic_on": "type", "polymorphic_identity": "interlocutor"}

    # Base primary key
    uuid: Mapped[UUID] = mapped_column(
        PSQL_UUID(as_uuid=True),
        primary_key=True,
        default=uuid4,
    )
    type: Mapped[str] = mapped_column(String, nullable=False)


class AIAgent(Interlocutor):
    __tablename__ = "ai_agents"

    # Attempt to override the primary key
    uuid: Mapped[UUID] = mapped_column(
        PSQL_UUID(as_uuid=True),
        ForeignKey("interlocutors.uuid"),
        primary_key=True,
        default=uuid4,
    )
    __mapper_args__ = {"polymorphic_identity": "ai_agent"}


# Inspect mapper's primary key for AIAgent
print(inspect(AIAgent).mapper.primary_key)
```
```
(Column('uuid', UUID(), table=<interlocutors>, primary_key=True, nullable=False, default=CallableColumnDefault(<function uuid4 at 0x104fc9080>)),)
```

As you can see, the `print` returns the Column that is defined in the `Interlocutor` table (rather than the one defined in `AIAgent`...

This inconsistency is breaking my [sqladmin](https://github.com/aminalaee/sqladmin) admin page when I try to create new elements of child tables.

**Comment by zzzeek** [comment]
the primary key for an `AIAgent` object is in fact the `Interlocutor.uuid` attribute.  `AIAgent.uuid` is constraint via foreign key to `Interlocutor.uuid` (and in fact you want to remove that default from `AIAgent.uuid`.

Now if you want to see the primary key **for the database table, not the Python class**, then you look at the `__table__` where you will see `ai_agents.uuid` as the primary key:   ``Interlocutor.__table__.c.uuid``

**Reply by AlbertMitjans** [reply]
Thank you for the quick reply! 🫶  Let me give you more context:

I have the following table:
```python3
class Company(Interlocutor):
    __tablename__ = "companies"

    uuid: Mapped[UUID] = mapped_column(
        PSQL_UUID(as_uuid=True),
        ForeignKey("interlocutors.uuid"),
        primary_key=True,
        default=uuid4,
    )

    ai_agent_id: Mapped[UUID] = mapped_column(
        PSQL_UUID(as_uuid=True),
        ForeignKey("ai_agents.uuid"),
        nullable=False,
    )

    ai_agent: Mapped[AIAgent] = relationship(
        "AIAgent", back_populates="company", foreign_keys=[ai_agent_id]
    )
```

When assigning a new `AIAgent` to a `Company` via the `sqladmin` admin panel, `sqladmin` does something like:
```python3
values = ... # fetches the UUID provided in the admin panel
pks = tuple(inspect(AIAgent).mapper.primary_key)
pk_values = {pk: value for pk, value in zip(pks, values)}

relationship = inspect(Company).mapper.relationships.get("ai_agent")
for fk, pk in relationship.local_remote_pairs:
    setattr(Company, fk.name, pk_values[pk])
```

You can see the code implementation here: https://github.com/aminalaee/sqladmin/blob/9b0874576b3bf64ab31bcede7b36868a5754a98c/sqladmin/_queries.py#L61-L73

The problem is that the `pk_values` dictionary contains `Interlocutor.uuid` as a key, while the `pk` from `relationship.local_remote_pairs` contains `AIAgent.uuid`...

Am I doing something wrong? If not, how could we fix the code from `sqladmin`?

**Reply by zzzeek** [reply]
that code is assuming that local_remote_pairs contains particular attributes inside of it which can't be assumed to match `mapper.primary_key` for the other side of things.   The "pk" value here is a `Column` and as mentioned previously you can key that via the `__table__`:


```py
pks = tuple(AIAgent.__table__.primary_key)
pk_values = {pk: value for pk, value in zip(pks, values)}
```

that will work in this exact example, however that's still assuming a particular foreign_key / primaryjoin setup for `Company.ai_agent`.    a primaryjoin / local_remote_pairs collection could have more than one column in it that aren't necessarily all primary key values anyway so this all depends on how generalized this tool is trying to be.   

since this seems to be reverse engineering things the ORM does, you probably have to use internals, and you can find equivalent candidates using `mapper._equivalent_columns` and this would get you there

```py
values = (uuid4(),)
pks = tuple(inspect(AIAgent).mapper.primary_key)
pk_values = {pk: value for pk, value in zip(pks, values)}

rel = inspect(Company).mapper.relationships.get("ai_agent")
c1 = Company()
for fk, pk in rel.local_remote_pairs:
    for pk_equiv in inspect(Company).mapper._equivalent_columns.get(pk, (pk, )):
        if pk_equiv in pk_values:
            setattr(c1, fk.name, pk_values[pk_equiv])
            break

```
